src/hotspot/share/opto/regmask.hpp

```diff
@@ -65,27 +68,286 @@ class RegMask {
+  //
+  // \_______________________________________________/ \_____________________________/
+  //                           |                                     |
+  //                         RM_UP                               RM_UP_EXT
+  // \_______________________________________________________________________________/
+  //                                         |
+  //                                     _rm_size
+
+  // Access word i in the register mask.
+  const uintptr_t& _rm_up(unsigned int i) const {
+    assert(orig_const || orig_ext_adr == &_RM_UP_EXT, "clone sanity check");
+    assert(i < _rm_size, "sanity");
+    if (i < _RM_SIZE) {
+      return _RM_UP[i];
+    } else {
+      assert(_RM_UP_EXT != nullptr, "sanity");
+      return _RM_UP_EXT[i - _RM_SIZE];
+    }
+  }
+
+  // Non-const version of the above.
+  uintptr_t& _rm_up(unsigned int i) {
+    assert(orig_ext_adr == &_RM_UP_EXT, "clone sanity check");
+    return const_cast<uintptr_t&>(const_cast<const RegMask*>(this)->_rm_up(i));
+  }
+
+  // The maximum word index
+  unsigned int _rm_max() const {
+    return _rm_size - 1U;
+  }
+
+  // Where to extend the register mask
+  Arena* _arena;
+
+  // Grow the register mask to ensure it can fit at least min_size words.
+  void _grow(unsigned int min_size, bool init = true) {
+    if (min_size > _rm_size) {
+      assert(min_size <= round_up_power_of_2(_RM_SIZE_MAX),
+             "unexpected register mask growth");
+      assert(_arena != nullptr, "register mask not growable");
+      min_size = round_up_power_of_2(min_size);
+      unsigned int old_size = _rm_size;
+      unsigned int old_ext_size = old_size - _RM_SIZE;
+      unsigned int new_ext_size = min_size - _RM_SIZE;
+      _rm_size = min_size;
+      if (_RM_UP_EXT == nullptr) {
+        assert(old_ext_size == 0, "sanity");
+        _RM_UP_EXT = NEW_ARENA_ARRAY(_arena, uintptr_t, new_ext_size);
+      } else {
+        assert(orig_ext_adr == &_RM_UP_EXT, "clone sanity check");
+        _RM_UP_EXT = REALLOC_ARENA_ARRAY(_arena, uintptr_t, _RM_UP_EXT,
+                                         old_ext_size, new_ext_size);
+      }
+      if (init) {
+        int fill = 0;
+        if (is_AllStack()) {
+          fill = 0xFF;
+          _hwm = _rm_max();
+        }
+        _set_range(old_size, fill, _rm_size - old_size);
+      }
+    }
+  }
+
+  // Make us a copy of src
+  void _copy(const RegMask& src) {
+    assert(_offset == src._offset, "offset mismatch");
+    _hwm = src._hwm;
+    _lwm = src._lwm;
+
+    // Copy base mask
+    memcpy(_RM_UP, src._RM_UP, sizeof(uintptr_t) * _RM_SIZE);
+    _all_stack = src._all_stack;
+
+    // Copy extension
+    if (src._RM_UP_EXT != nullptr) {
+      assert(src._rm_size > _RM_SIZE, "sanity");
+      assert(orig_ext_adr == &_RM_UP_EXT, "clone sanity check");
+      _grow(src._rm_size, false);
+      memcpy(_RM_UP_EXT, src._RM_UP_EXT,
+             sizeof(uintptr_t) * (src._rm_size - _RM_SIZE));
```

Comment:
This code is not very well covered by current tests, please consider adding some tests to `test_regmask.cpp` to exercise it.

Reply:
Now added!